tests/test_io.py

```diff
@@ -4,29 +4,141 @@
+    message_parameters = message_parameters_dict[message_format]
+    model_name = message_parameters["model_name"]
+    expected_model = message_parameters["expected_model"]
+
+    # test valid formats
+    with patch(f"hop.models.{model_name}", MagicMock()):
+        test_model = io.Deserializer.deserialize(message)
+
+    # verify the message is classified properly
+    assert isinstance(test_model, expected_model)
+
+
+def test_stream_read(circular_msg, circular_text):
+    with patch("hop.io._Consumer", MagicMock()) as mock_consumer:
+        mock_consumer.stream.return_value = [{'hey', 'you'}]
+
+        broker_url1 = "kafka://hostname:port/gcn"
+        broker_url2 = "kafka://hey@hostname:port/gcn"
+        start_at = io.StartPosition.EARLIEST
+        persist = False
+
+        stream = io.Stream(persist=persist)
+
+        with stream.open(broker_url1, "r", start_at=start_at) as s:
+            for msg in s:
+                continue
+
+        with stream.open(broker_url2, "r", start_at=start_at) as s:
+            for msg in s:
+                continue
+
+
+def test_stream_write(circular_msg, circular_text):
+    with patch("hop.io._Producer", MagicMock()) as mock_producer:
+        mock_producer.write = Mock()
+
+        broker_url = "kafka://localhost:port/gcn"
+        auth = Auth("user", "password")
+        start_at = io.StartPosition.EARLIEST
+        persist = False
+
+        stream = io.Stream()
+
+        # verify only 1 topic is allowed in write-mode
+        with pytest.raises(ValueError):
+            stream.open("kafka://localhost:9092/topic1,topic2", "w")
+
+        # check various warnings when certain settings are set
+        with pytest.warns(UserWarning):
+            stream.open("kafka://group@localhost:9092/topic1", "w")
+        with pytest.warns(UserWarning):
+            stream.open(broker_url, "w", start_at=start_at)
+        with pytest.warns(UserWarning):
+            stream.open(broker_url, "w", persist=persist)
+
+        with stream.open(broker_url, "w", auth=auth) as s:
             s.write(circular_msg)
 
-        # verify GCN was processed
-        mock_stream.assert_called_with(
-            broker_url, "w", format=format, start_at=start_at, config=config
-        )
+
+def test_stream_open():
+    stream = io.Stream()
+
+    # verify only read/writes are allowed
+    with pytest.raises(ValueError):
+        stream.open("kafka://localhost:9092/topic1", "q")
+
+
+def test_unpack(circular_msg, circular_text):
+    wrapped_msg = {"format": "circular", "content": circular_msg}
+
+    kafka_msg = MagicMock()
+    kafka_msg.value.return_value = json.dumps(wrapped_msg).encode("utf-8")
+
+    unpacked_msg = io._Consumer.unpack(kafka_msg)
+
+
+def test_pack(circular_msg, circular_text):
+    # message class
+    circular = GCNCircular(**circular_msg)
+    packed_msg = io._Producer.pack(circular)
+
+    # unstructured message
+    message = {"hey": "you"}
+    packed = io._Producer.pack(message)
```

Comment:
Maybe add a round-trip test for each format: pack a message and then unpack it; it should be unchanged.

Reply:
That's a good idea, I'll go add that in when I get a chance.

Reply:
This has now been added.